lib/ast-utils.js

```diff
@@ -0,0 +1,174 @@
+     */
+    if (node.type === "Program") {
+        if (node.body.length === 0) {
+            leadingComments = node.comments;
+        }
+    }
+
+    return {
+        leading: leadingComments,
+        trailing: trailingComments
+    };
+}
+
+/**
+ * Retrieves the JSDoc comment for a given node.
+ * @param {ASTNode} node The AST node to get the comment for.
+ * @returns {ASTNode} The BlockComment node containing the JSDoc for the
+ *      given node or null if not found.
+ * @public
+ */
+function getJSDocComment(node) {
+
+    var parent = node.parent,
+        line = node.loc.start.line;
+
+    switch (node.type) {
+        case "FunctionDeclaration":
+            if (looksLikeExport(parent)) {
+                return findJSDocComment(parent.leadingComments, line);
+            } else {
+                return findJSDocComment(node.leadingComments, line);
+            }
+            break;
+
+        case "ArrowFunctionExpression":
+        case "FunctionExpression":
+
+            if (parent.type !== "CallExpression") {
+                while (parent && !parent.leadingComments && !/Function/.test(parent.type)) {
+                    parent = parent.parent;
+                }
+
+                return parent && (parent.type !== "FunctionDeclaration") ? findJSDocComment(parent.leadingComments, line) : null;
+            }
+
+        // falls through
+
+        default:
+            return null;
+    }
+}
+
+/**
+ * Determines whether two adjacent tokens are have whitespace between them.
+ * @param {Object} left - The left token object.
+ * @param {Object} right - The right token object.
+ * @returns {boolean} Whether or not there is space between the tokens.
+ * @public
+ */
+function isTokenSpaced(left, right) {
+    return left.range[1] < right.range[0];
+}
+
+/**
+ * Determines whether two adjacent tokens are on the same line.
+ * @param {Object} left - The left token object.
+ * @param {Object} right - The right token object.
+ * @returns {boolean} Whether or not the tokens are on the same line.
+ * @public
+ */
+function isTokenOnSameLine(left, right) {
+    return left.loc.end.line === right.loc.start.line;
+}
+
+/**
+ * Checks whether or not a node is `null` or `undefined`.
+ * @param {ASTNode} node - A node to check.
+ * @returns {boolean} Whether or not the node is a `null` or `undefined`.
+ * @public
+ */
+function isNullOrUndefined(node) {
+    return (
+        (node.type === "Literal" && node.value === null) ||
+        (node.type === "Identifier" && node.name === "undefined") ||
+        (node.type === "UnaryExpression" && node.operator === "void")
+    );
+}
+
+/**
+ * Checks reference if is non initializer and writable.
+ * @param {Reference} reference - A reference to check.
+ * @param {int} index - The index of the reference in the references.
+ * @param {Reference[]} references - The array that the reference belongs to.
+ * @returns {boolean} Success/Failure
+ */
+function checkReference(reference, index, references) {
```

Comment:
I want this function to keep private because its argument list is mysterious.

I suggest such as below instead.

```js
/**
 * Checks reference if is non initializer and writable.
 * @param {Reference} reference - A reference to check.
 * @param {int} index - The index of the reference in the references.
 * @param {Reference[]} references - The array that the reference belongs to.
 * @returns {boolean} Success/Failure
 * @private
 */
function isModifyingReference(reference, index, references) {
    var identifier = reference.identifier;

    return (identifier != null &&
        reference.init === false &&
        reference.isWrite() &&
        // Destructuring assignments can have multiple default value,
        // so possibly there are multiple writeable references for the same identifier.
        (index === 0 || references[index - 1].identifier !== identifier)
    );
}


/**
 * Gets references which are non initializer and writable.
 * @param {Reference[]} references - An array of references.
 * @returns {Reference[]} An array of only references which are non initializer and writable.
 */
function getModifyingReferences(references) {
  return references.filter(isModifyingReference);
}

module.exports = {
    ...,
    getModifyingReferences: getModifyingReferences
};
```